File: from-1601-to-1700/1609-even-odd-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func isEvenOddTree(root *TreeNode) bool {
    var n, prev, val int
    var node *TreeNode
    level := 0
    q := []*TreeNode{root}

    for len(q) > 0 {
        n = len(q)
        prev = q[0].Val
        for i := 0; i < n; i ++ {
            node = q[0]
            q = q[1:]
            val = node.Val
            if level % 2 == 0 && val % 2 == 0 || level % 2 != 0 && val % 2 != 0 {
                return false
            }
            if i > 0 {
                if level % 2 == 0 && val <= prev || level % 2 != 0 && val >= prev{
                    return false
                }
                prev = val
            }
            if node.Left != nil {
                q = append(q, node.Left)
            }
            if node.Right != nil {
                q = append(q, node.Right)
            }
        }
        level += 1
    }
    return true
}

File: from-1601-to-1700/1609-even-odd-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        def strictly_inc(vals: list[int]) -> bool:
            for i in range(1, len(vals)):
                if vals[i] <= vals[i-1]:
                    return False
            return True
        
        level = 0
        q = deque()
        q.append(root)
        
        while q:
            vals = []
            for _ in range(len(q)):
                node = q.popleft()
                vals.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
                    
            if not strictly_inc(vals if level & 1 == 0 else vals[::-1]):
                return False
            
            cond = (lambda x: x % 2 == 1) if level & 1 else (lambda x: x % 2 == 0)
            if len(list(filter(cond, vals))) != 0:
                return False
            
            level += 1
            
        return True